### PR TITLE
feat: add alter columns for names and nullability

### PR DESCRIPTION
**Pull request by wjones127**
Initial pass on `alter_columns()` API. This allows renaming columns and making them nullable. A future PR will allow casting the type of column.

### CODE REVIEW

File: python/src/dataset.rs [hunk 1]
```diff
@@ -629,6 +630,40 @@ impl Dataset {
         Ok(PyArrowType(Box::new(LanceReader::from_stream(stream))))
     }
 
+    fn alter_columns(&mut self, alterations: &PyList) -> PyResult<()> {
+        let alterations = alterations
+            .iter()
+            .map(|obj| {
+                let obj = PyAny::downcast::<PyDict>(obj)?;
```
Comment (westonpace) [inline python/src/dataset.rs:637]: Can you not just say `obj.downcast::<PyDict>()`?

File: rust/lance/src/dataset.rs [hunk 2]
```diff
@@ -1657,6 +1691,69 @@ impl Dataset {
         Ok(fragments)
     }
 
+    /// Modify columns in the dataset, changing their name, type, or nullability.
+    ///
+    /// If a column has an index, it's index will be preserved or transformed to
+    /// the new type as part of the operation.
```
Comment (westonpace) [inline rust/lance/src/dataset.rs:1697]: Index transformation is only future right?  Once we support data type changes?

File: python/python/lance/dataset.py [hunk 3]
```diff
@@ -610,6 +610,43 @@ def join(
...
+                The column path to alter. For a top-level column, this is the name.
+                For a nested column, this is the dot-separated path, e.g. "a.b.c".
+            - "name": str, optional
+                The new name of the column. If not specified, the column name is
+                not changed.
+            - "nullable": bool, optional
+                Whether the column should be nullable. If not specified, the column
+                nullability is not changed. Only non-nullable columns can be changed
+                to nullable. Currently, you cannot change a nullable column to
+                non-nullable.
```
Comment (westonpace) [inline python/python/lance/dataset.py:630]: If the column is already nullable, or already has the desired name, is it a no-op or a failure?  From the code it looks like a no-op correct?

**Review by westonpace — APPROVED** [review]
Looks good!

File: python/python/lance/dataset.py (hunk 3 above)
Comment (wjones127) [reply python/python/lance/dataset.py:630]: Yes, it should be a no-op.

File: rust/lance/src/dataset.rs (hunk 2 above)
Comment (wjones127) [reply rust/lance/src/dataset.rs:1697]: Yeah I should remove that for now.

**Pull request merged**